crates/vessel/src/channel/db.rs

```diff
@@ -0,0 +1,562 @@
+use std::{collections::HashSet, time::Duration};
+
+use chrono::{DateTime, Utc};
+use color_eyre::eyre::{Result, eyre};
+use futures_util::TryStreamExt;
+use service::database::Transaction;
+use sqlx::{FromRow, SqliteConnection};
+use tokio::time;
+use tracing::debug;
+
+use crate::channel::version::{Entry, HistoryIdentifier, Identifier, TagIdentifier, Version};
+
+/// Lookup if a package name already exists in the provided channel version
+pub async fn lookup_entry(
+    conn: &mut SqliteConnection,
+    channel: &str,
+    version: &Version,
+    name: &str,
+) -> sqlx::Result<Option<Entry>> {
+    let Some(history) = find_related_history(conn, channel, version).await? else {
+        return Ok(None);
+    };
+
+    sqlx::query_as(
+        "
+        SELECT
+          package_id,
+          name,
+          arch,
+          source_id,
+          source_version,
+          source_release,
+          build_release
+        FROM
+          channel_version_entry
+        WHERE
+          channel_version_id = ?
+          AND name = ?;
+        ",
+    )
+    .bind(history.channel_version_id)
+    .bind(name)
+    .fetch_optional(&mut *conn)
+    .await
+}
+
+/// List entries for the provided channel version
+pub async fn entries(conn: &mut SqliteConnection, channel: &str, version: &Version) -> sqlx::Result<Vec<Entry>> {
+    let Some(history) = find_related_history(conn, channel, version).await? else {
+        return Ok(vec![]);
+    };
+
+    sqlx::query_as(
+        "
+        SELECT
+          package_id,
+          name,
+          arch,
+          source_id,
+          source_version,
+          source_release,
+          build_release
+        FROM
+          channel_version_entry
+        WHERE
+          channel_version_id = ?;
+        ",
+    )
+    .bind(history.channel_version_id)
+    .fetch_all(&mut *conn)
```

Comment:
You don't need the explicit reborrow, right?

```suggestion
    .fetch_all(conn)
```